### DISCUSSION TITLE
The second traffic selector is not selected for remote IPsec communication

### DISCUSSION THREAD

**Discussion by yizhao1** [strongswan/strongswan#1129] [General]
Hi,

I have a question about the traffic selector. In my environment, I found the IPsec does not work for remote communication of the second traffic selector on the responder. But it works when the traffic selector is set to the first.

Here is my environment:

There are 2 hosts: Host A (192.168.0.105) and Host B (192.168.0.109)

If Host B is in the second traffic selector (child_2) in swanctl.conf on Host A. I can not ping Host A from Host B. But I can ping Host B from Host A. After this, I can ping Host A from Host B. There is no such issue if Host B is in the first traffic selector (child_1).

swanctl.conf on Host A:

connections {

   conn {
      version = 1
      local_addrs = %any
      remote_addrs = %any
      proposals = 3des-sha1-modp1024

      local {
         auth = psk
      }
      remote {
         auth = psk
      }

      children {
         child_1 {
            remote_ts = 192.168.0.100
            esp_proposals = 3des-sha1
            mode = transport
            start_action = trap
         }
         child_2 {
            remote_ts = 192.168.0.109
            esp_proposals = 3des-sha1
            mode = transport
            start_action = trap
         }
      }
   }
}

secrets {
      ike {
         secret = "12345678"
      }
}

swanctl.conf on Host B:

connections {

	conn {
		version = 1
		local_addrs = %any
		remote_addrs = %any
		proposals = 3des-sha1-modp1024

		local {
			auth = psk
		}
		remote {
			auth = psk
		}

		children {
			child_1 {
				remote_ts = 192.168.0.105
				esp_proposals = 3des-sha1
				mode = transport
				start_action = trap
			}
		}
	}
}

secrets {
	ike {
		secret = "12345678"
	}
}

Seems this is introduced by commit:
https://github.com/strongswan/strongswan/commit/da82786b2d8cef68ca6462bf7898a6b19c0b4608

Is there any way to solve this issue? Thanks.

**Comment by Thermi** [comment]
Does it work if you just define a second connection with that child section?

**Reply by yizhao1** [reply]
Yes. It works if there is only one selector (child_2).

**Comment by tobiasbrunner** [comment]
You can't use IP addresses different from the IKE endpoints for transport mode SAs. Either create separate IKE/IPsec SAs for each distinct address pair, use tunnel mode, or look into BEET mode (not standardized and strongSwan only implements it for IKEv2).

Also, please don't use IKEv1 between two strongSwan instances.

**Reply by yizhao1** [reply]
I have set local_addrs/remote_addrs to %any. Does this not work ? I'm migrating from old strongswan version to new version 5.9 and this setting used to work in 5.3.

**Reply by tobiasbrunner** [reply]
I overlooked that (incorrectly formatted config). I don't see a reason why it shouldn't work (you also haven't provided any prove, like logs etc., that it doesn't). Why did you configure a simple host-to-host connection like this, though?

**Reply by yizhao1** [reply]
When Host B (192.168.0.109) is in the second traffic selector (child_2) in swanctl.conf on Host A (192.168.0.105).
Ping Host A from Host B.
I got the following logs in Host A:

received packet: from 192.168.0.109[500] to 192.168.0.105[500]
parsing header of message
parsing HEADER payload, 332 bytes left 
[snip]
parsing HEADER payload finished
parsed a IKE_SA_INIT request header
waiting for data on sockets
checkout IKEv2 SA by message with SPIs c804ebfbbde98e1a_i 0000000000000000_r
created IKE_SA (unnamed)[1]
received packet: from 192.168.0.109[500] to 192.168.0.105[500] (332 bytes)
parsing body of message, first payload is SECURITY_ASSOCIATION
starting parsing a SECURITY_ASSOCIATION payload
parsing SECURITY_ASSOCIATION payload, 304 bytes left
[snip]
authentication of '192.168.0.109' with pre-shared key successful
peer supports MOBIKE
no IDr configured, fall back on IP address
added payload of type ID_RESPONDER to message
authentication of '192.168.0.105' (myself) with pre-shared key
successfully created shared key MAC
added payload of type AUTH to message
IKE_SA conn[1] established between 192.168.0.105[192.168.0.105]...192.168.0.109[192.168.0.109]
IKE_SA conn[1] state change: CONNECTING => ESTABLISHED
next event in 29s 957ms, waiting
scheduling rekeying in 13321s
next event in 29s 957ms, waiting
maximum IKE_SA lifetime 14761s
looking for a child config for 192.168.0.105/32[udp/1025] 192.168.0.105/32 === 192.168.0.109/32[udp/56269] 192.168.0.109/32
proposing traffic selectors for us:
 192.168.0.105/32
proposing traffic selectors for other:
 192.168.0.109/32
  candidate "child_1" with prio 7+7
proposing traffic selectors for us:
 192.168.0.105/32
proposing traffic selectors for other:
 192.168.0.109/32
  candidate "child_2" with prio 7+7
found matching child config "child_1" with prio 14
selecting proposal:
  proposal matches
received proposals: ESP:3DES_CBC/HMAC_SHA1_96/NO_EXT_SEQ
configured proposals: ESP:3DES_CBC/HMAC_SHA1_96/NO_EXT_SEQ
selected proposal: ESP:3DES_CBC/HMAC_SHA1_96/NO_EXT_SEQ
got SPI c7a8a797
selecting traffic selectors for us:
 config: 192.168.0.105/32, received: 192.168.0.105/32[udp/1025] => match: 192.168.0.105/32[udp/1025]
 config: 192.168.0.105/32, received: 192.168.0.105/32 => match: 192.168.0.105/32
selecting traffic selectors for other:
 config: 192.168.0.100/32, received: 192.168.0.109/32[udp/56269] => no match
 config: 192.168.0.100/32, received: 192.168.0.109/32 => no match
no acceptable traffic selectors found
added payload of type NOTIFY to message
failed to establish CHILD_SA, keeping IKE_SA
CHILD_SA child_1{3} state change: CREATED => DESTROYING


Seems there is no matching traffic selector.


When move Host B (192.168.0.109) to the first traffic selector.
Ping Host A from Host B.
I got the following logs in Host A:

received packet: from 192.168.0.109[500] to 192.168.0.105[500]
parsing header of message
parsing HEADER payload, 332 bytes left 
[snip]
parsing HEADER payload finished
parsed a IKE_SA_INIT request header
waiting for data on sockets
checkout IKEv2 SA by message with SPIs cbd0b137b28f9df0_i 0000000000000000_r
created IKE_SA (unnamed)[1]
received packet: from 192.168.0.109[500] to 192.168.0.105[500] (332 bytes)
parsing body of message, first payload is SECURITY_ASSOCIATION
starting parsing a SECURITY_ASSOCIATION payload
parsing SECURITY_ASSOCIATION payload, 304 bytes left
[snip]
authentication of '192.168.0.109' with pre-shared key successful
peer supports MOBIKE
no IDr configured, fall back on IP address
added payload of type ID_RESPONDER to message
authentication of '192.168.0.105' (myself) with pre-shared key
successfully created shared key MAC
added payload of type AUTH to message
IKE_SA conn[1] established between 192.168.0.105[192.168.0.105]...192.168.0.109[192.168.0.109]
IKE_SA conn[1] state change: CONNECTING => ESTABLISHED
next event in 29s 965ms, waiting
scheduling rekeying in 14087s
next event in 29s 965ms, waiting
maximum IKE_SA lifetime 15527s
looking for a child config for 192.168.0.105/32[udp/1025] 192.168.0.105/32 === 192.168.0.109/32[udp/56939] 192.168.0.109/32
proposing traffic selectors for us:
 192.168.0.105/32
proposing traffic selectors for other:
 192.168.0.109/32
  candidate "child_1" with prio 7+7
proposing traffic selectors for us:
 192.168.0.105/32
proposing traffic selectors for other:
 192.168.0.109/32
  candidate "child_2" with prio 7+7
found matching child config "child_1" with prio 14
selecting proposal:
  proposal matches
received proposals: ESP:3DES_CBC/HMAC_SHA1_96/NO_EXT_SEQ
configured proposals: ESP:3DES_CBC/HMAC_SHA1_96/NO_EXT_SEQ
selected proposal: ESP:3DES_CBC/HMAC_SHA1_96/NO_EXT_SEQ
got SPI c60252aa
selecting traffic selectors for us:
 config: 192.168.0.105/32, received: 192.168.0.105/32[udp/1025] => match: 192.168.0.105/32[udp/1025]
 config: 192.168.0.105/32, received: 192.168.0.105/32 => match: 192.168.0.105/32
selecting traffic selectors for other:
 config: 192.168.0.109/32, received: 192.168.0.109/32[udp/56939] => match: 192.168.0.109/32[udp/56939]
 config: 192.168.0.109/32, received: 192.168.0.109/32 => match: 192.168.0.109/32
CHILD_SA child_1{3} state change: CREATED => INSTALLING
  using 3DES_CBC for encryption
  using HMAC_SHA1_96 for integrity
adding inbound ESP SA
  SPI 0xc60252aa, src 192.168.0.109 dst 192.168.0.105
adding SAD entry with SPI c60252aa and reqid {3}
  using encryption algorithm 3DES_CBC with key size 192
  using integrity algorithm HMAC_SHA1_96 with key size 160
  using replay window of 32 packets
  HW offload: no
adding outbound ESP SA
  SPI 0xc5e28613, src 192.168.0.105 dst 192.168.0.109
adding SAD entry with SPI c5e28613 and reqid {3}
  using encryption algorithm 3DES_CBC with key size 192
  using integrity algorithm HMAC_SHA1_96 with key size 160
  using replay window of 0 packets
  HW offload: no
adding policy 192.168.0.109/32 === 192.168.0.105/32 in [priority 367231, refcount 1]
adding policy 192.168.0.105/32 === 192.168.0.109/32 out [priority 367231, refcount 1]
CHILD_SA child_1{3} established with SPIs c60252aa_i c5e28613_o and TS 192.168.0.105/32 === 192.168.0.109/32
CHILD_SA child_1{3} state change: INSTALLING => INSTALLED
added payload of type SECURITY_ASSOCIATION to message
added payload of type TS_INITIATOR to message
added payload of type TS_RESPONDER to message

**Reply by tobiasbrunner** [reply]
Thanks for the logs. You're right, this is a bug. And the commit you pointed out is, in fact, the culprit. When I wrote it, I missed that the config selection process uses a `get_traffic_selectors()` call that's basically exactly like the one used to generate traffic selectors as initiator (i.e. with hosts but no supplied TS). I pushed a possible fix to the _1143-transport-ts_ branch.

**Reply by yizhao1** [reply]
Thanks. I have tested the patch. It works for me.
